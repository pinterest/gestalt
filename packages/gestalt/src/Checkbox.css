:root {
  --red: #bd081c;
  --size-sm: 16px;
  --size-md: 24px;
}

.border {
  border-color: #8e8e8e;
}

.borderError {
  border-color: #e3780c;
}

.borderDark {
  border-color: #333;
}

.check {
  composes: block from "./Layout.css";
  composes: borderBox from "./Layout.css";
  composes: flex from "./Layout.css";
  composes: itemsCenter from "./Layout.css";
  composes: justifyCenter from "./Layout.css";
  border-style: solid;
  border-width: 1px;
}

.checkEnabled {
  composes: pointer from "./Cursor.css";
}

.checkFocused {
  composes: accessibilityOutlineFocus from "./Focus.css";
}

.checkSm {
  border-radius: 3px;
  height: var(--size-sm);
  width: var(--size-sm);
}

.checkMd {
  border-radius: 4px;
  height: var(--size-md);
  width: var(--size-md);
}

.input {
  composes: absolute from "./Layout.css";
  composes: m0 from "./Whitespace.css";
  appearance: none;
  opacity: 0;
  outline: 0;
}

.inputEnabled {
  composes: pointer from "./Cursor.css";
}

.inputSm {
  height: var(--size-sm);
  width: var(--size-sm);
}

.inputMd {
  height: var(--size-md);
  width: var(--size-md);
}
