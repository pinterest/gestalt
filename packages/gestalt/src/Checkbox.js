// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import colors from './Colors.css';
import styles from './Checkbox.css';
import Box from './Box.js';
import Icon from './Icon.js';

type Props = {|
  checked?: boolean,
  disabled?: boolean,
  hasError?: boolean,
  id: string,
  indeterminate?: boolean,
  name?: string,
  onChange: ({ event: SyntheticInputEvent<>, checked: boolean }) => void,
  onClick?: ({
    event: SyntheticInputEvent<HTMLInputElement>,
    checked: boolean,
  }) => void,
  size?: 'sm' | 'md',
|};

type State = {|
  focused: boolean,
|};

export default class Checkbox extends React.Component<Props, State> {
  static propTypes = {
    checked: PropTypes.bool,
    disabled: PropTypes.bool,
    hasError: PropTypes.bool,
    id: PropTypes.string.isRequired,
    indeterminate: PropTypes.bool,
    name: PropTypes.string,
    onChange: PropTypes.func.isRequired,
    onClick: PropTypes.func,
    size: PropTypes.oneOf(['sm', 'md']),
  };

  static defaultProps = {
    checked: false,
    disabled: false,
    hasError: false,
    indeterminate: false,
    size: 'md',
  };

  state = {
    focused: false,
  };

  componentDidMount() {
    if (this.props.indeterminate) {
      this.setIndeterminate(!!this.props.indeterminate);
    }
  }

  componentDidUpdate(previousProps: Props) {
    if (previousProps.indeterminate !== this.props.indeterminate) {
      this.setIndeterminate(!!this.props.indeterminate);
    }
  }

  setIndeterminate(indeterminate: boolean) {
    if (this.input) {
      this.input.indeterminate = indeterminate;
    }
  }

  handleChange = (event: SyntheticInputEvent<>) => {
    const { checked } = event.target;
    this.props.onChange({ event, checked });
  };

  handleClick = (event: SyntheticInputEvent<HTMLInputElement>) => {
    const { onClick } = this.props;
    if (onClick) {
      const { checked } = event.currentTarget;
      onClick({ event, checked });
    }
  };

  handleBlur = () => this.setState({ focused: false });

  handleFocus = () => this.setState({ focused: true });

  input: ?HTMLInputElement;

  render() {
    const {
      checked,
      disabled,
      hasError,
      id,
      indeterminate,
      name,
      size,
    } = this.props;

    let borderStyle = styles.border;
    if (!disabled && (checked || indeterminate)) {
      borderStyle = styles.borderDark;
    } else if (hasError) {
      borderStyle = styles.borderError;
    }

    return (
      <Box position="relative">
        <input
          checked={checked}
          className={classnames(styles.input, {
            [styles.inputEnabled]: !disabled,
            [styles.indeterminate]: indeterminate,
            [styles.inputSm]: size === 'sm',
            [styles.inputMd]: size === 'md',
          })}
          disabled={disabled}
          id={id}
          name={name}
          onBlur={this.handleBlur}
          onChange={this.handleChange}
          onClick={this.handleClick}
          onFocus={this.handleFocus}
          ref={el => {
            this.input = el;
          }}
          type="checkbox"
        />
        <div
          className={classnames(
            borderStyle,
            styles.check,
            // eslint-disable-next-line no-nested-ternary
            disabled
              ? checked || indeterminate
                ? colors.grayBg
                : colors.lightGrayBg
              : checked || indeterminate
                ? colors.darkGrayBg
                : colors.whiteBg,
            {
              [styles.checkEnabled]: !disabled,
              [styles.checkFocused]: this.state.focused,
              [styles.checkMd]: size === 'md',
              [styles.checkSm]: size === 'sm',
            }
          )}
        >
          {(checked || indeterminate) && (
            <Icon
              accessibilityLabel=""
              color="white"
              icon={indeterminate ? 'dash' : 'check'}
              size={size === 'sm' ? 8 : 12}
            />
          )}
        </div>
      </Box>
    );
  }
}
