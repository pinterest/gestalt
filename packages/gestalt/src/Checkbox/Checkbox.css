:root {
  --red: #bd081c;
  --size-sm: 16px;
  --size-md: 24px;
}

.check {
  composes: block from "../Layout.css";
  composes: borderBox from "../Layout.css";
  composes: flex from "../Layout.css";
  composes: itemsCenter from "../Layout.css";
  composes: justifyCenter from "../Layout.css";
  border-style: solid;
  border-width: 1px;
}

.checkEnabled {
  composes: pointer from "../Cursor.css";
}

.checkDarkGray {
  composes: darkGrayBg from "../Colors.css";
  border-color: #555;
}

.checkGray {
  composes: grayBg from "../Colors.css";
  border-color: #b5b5b5;
}

.checkLightGray {
  composes: lightGrayBg from "../Colors.css";
  border-color: #b5b5b5;
}

.checkWhite {
  composes: whiteBg from "../Colors.css";
  border-color: #b5b5b5;
}

.checkFocused {
  composes: accessibilityOutlineFocus from "../Focus.css";
}

.checkSm {
  border-radius: 3px;
  height: var(--size-sm);
  width: var(--size-sm);
}

.checkMd {
  border-radius: 4px;
  height: var(--size-md);
  width: var(--size-md);
}

.input {
  composes: absolute from "../Layout.css";
  composes: m0 from "../Whitespace.css";
  appearance: none;
  opacity: 0;
  outline: 0;
}

.inputEnabled {
  composes: pointer from "../Cursor.css";
}

.inputSm {
  height: var(--size-sm);
  width: var(--size-sm);
}

.inputMd {
  height: var(--size-md);
  width: var(--size-md);
}
