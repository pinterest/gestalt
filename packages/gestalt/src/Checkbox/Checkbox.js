// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import styles from './Checkbox.css';
import Box from '../Box/Box';
import Icon from '../Icon/Icon';

type Props = {|
  checked?: boolean,
  disabled?: boolean,
  id: string,
  indeterminate?: boolean,
  name?: string,
  onChange: ({ event: SyntheticInputEvent<>, checked: boolean }) => void,
  size?: 'sm' | 'md',
|};

type State = {|
  focused: boolean,
|};

export default class Checkbox extends React.Component<Props, State> {
  static propTypes = {
    checked: PropTypes.bool,
    disabled: PropTypes.bool,
    id: PropTypes.string.isRequired,
    indeterminate: PropTypes.bool,
    name: PropTypes.string,
    onChange: PropTypes.func.isRequired,
    size: PropTypes.oneOf(['sm', 'md']),
  };

  static defaultProps = {
    checked: false,
    disabled: false,
    indeterminate: false,
    size: 'md',
  };

  state: State = {
    focused: false,
  };

  componentDidMount() {
    if (this.props.indeterminate) {
      this.setIndeterminate(!!this.props.indeterminate);
    }
  }

  componentDidUpdate(previousProps: Props) {
    if (previousProps.indeterminate !== this.props.indeterminate) {
      this.setIndeterminate(!!this.props.indeterminate);
    }
  }

  setIndeterminate(indeterminate: boolean) {
    if (this.input) {
      this.input.indeterminate = indeterminate;
    }
  }

  input: ?HTMLInputElement;

  handleChange = (event: SyntheticInputEvent<>) => {
    const { checked } = event.target;
    this.props.onChange({ event, checked });
  };

  handleBlur = () => this.setState({ focused: false });
  handleFocus = () => this.setState({ focused: true });

  render() {
    const { checked, disabled, id, indeterminate, name, size } = this.props;
    return (
      <Box position="relative">
        <input
          checked={checked}
          className={classnames(styles.input, {
            [styles.inputEnabled]: !disabled,
            [styles.indeterminate]: indeterminate,
            [styles.inputSm]: size === 'sm',
            [styles.inputMd]: size === 'md',
          })}
          disabled={disabled}
          id={id}
          name={name}
          onBlur={this.handleBlur}
          onChange={this.handleChange}
          onFocus={this.handleFocus}
          ref={el => {
            this.input = el;
          }}
          type="checkbox"
        />
        <div
          className={classnames(
            styles.check,
            // eslint-disable-next-line no-nested-ternary
            disabled
              ? checked || indeterminate
                ? styles.checkGray
                : styles.checkLightGray
              : checked || indeterminate
                ? styles.checkDarkGray
                : styles.checkWhite,
            {
              [styles.checkEnabled]: !disabled,
              [styles.checkFocused]: this.state.focused,
              [styles.checkMd]: size === 'md',
              [styles.checkSm]: size === 'sm',
            }
          )}
        >
          {(checked || indeterminate) && (
            <Icon
              accessibilityLabel=""
              color="white"
              icon={indeterminate ? 'dash' : 'check'}
              size={size === 'sm' ? 8 : 12}
            />
          )}
        </div>
      </Box>
    );
  }
}
