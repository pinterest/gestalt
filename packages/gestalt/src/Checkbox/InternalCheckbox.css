:root {
  --g-size-sm: var(--space-400);
  --g-size-md: var(--space-600);
}

.sizeSm {
  height: var(--g-size-sm);
  width: var(--g-size-sm);
}

.sizeMd {
  height: var(--g-size-md);
  width: var(--g-size-md);
}

.enabled {
  background-color: var(--color-background-checkbox-unchecked-default);
}

.disabled {
  background-color: var(--color-background-checkbox-disabled);
}

.checked {
  background-color: var(--color-background-checkbox-checked-default);
}

.errorChecked {
  background-color: var(--color-background-checkbox-checked-error);
}

.error {
  background-color: var(--color-background-checkbox-unchecked-error);
}

.border {
  border-color: var(--color-border-checkbox-unchecked-default);
}

.borderSelected {
  border-color: var(--color-border-checkbox-checked-default);
}

.borderDisabled {
  border-color: var(--color-border-checkbox-disabled);
}

.borderErrorChecked {
  border-color: var(--color-border-checkbox-checked-error);
}

.borderErrorUnchecked {
  border-color: var(--color-border-checkbox-unchecked-error);
}

.borderHovered {
  border-color: var(--color-border-checkbox-unchecked-hover);
}

.borderHoveredError {
  border-color: var(--sema-color-hover-border-error);
}

.borderPressed {
  border-color: var(--sema-color-pressed-border-interactive);
}

.borderPressedError {
  border-color: var(--sema-color-pressed-border-error);
}

.hovered {
  background-color: var(--sema-color-hover-background-selected);
  border-color: var(--sema-color-hover-background-selected);
}

.focus {
  border-color: var(--sema-color-border-focus-inner-default);
  border-style: solid;
  border-width: 2px;
  outline: 2px solid var(--sema-color-border-focus-outer-default);
  outline-offset: 0;
}

.hoveredError {
  background-color: var(--sema-color-hover-background-error-weak);
  border-color: var(--sema-color-hover-background-error-weak);
}

.pressed {
  background-color: var(--sema-color-pressed-background-selected);
  border-color: var(--sema-color-pressed-background-selected);
}

.pressedError {
  background-color: var(--sema-color-pressed-background-error-weak);
  border-color: var(--sema-color-pressed-background-error-weak);
}

.borderRadiusSm {
  composes: rounding1 from "../Borders.css";
}

.borderRadiusMd {
  composes: rounding2 from "../Borders.css";
}

.vrBorderRadiusSm {
  border-radius: var(--sema-rounding-xs);
}

.vrBorderRadiusMd {
  border-radius: var(--sema-rounding-sm);
}

.check {
  composes: block from "../Layout.css";
  composes: borderBox from "../Layout.css";
  composes: flex from "../Layout.css";
  composes: xsItemsCenter from "../Layout.css";
  composes: justifyCenter from "../Layout.css";
  composes: solid from "../Borders.css";
}

.sizeLg {
  height: var(--space-600);
  width: var(--space-600);
}

.thinBorder {
  composes: sizeSm from "../Borders.css";
}

.thickBorder {
  composes: sizeLg from "../Borders.css";
}

.inputEnabled {
  composes: pointer from "../Cursor.css";
}

.readOnly {
  pointer-events: none;
}

.input {
  composes: absolute from "../Layout.css";
  composes: m0 from "../Whitespace.css";
  appearance: none;
  opacity: var(--opacity-0);
  outline: 0;
}
