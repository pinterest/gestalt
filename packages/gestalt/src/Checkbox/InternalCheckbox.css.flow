// @flow strict
/* This file is automatically generated by css-modules-flow-types */
declare module.exports: {|
  +'border': string,
  +'borderDarkGray': string,
  +'borderDisabled': string,
  +'borderError': string,
  +'borderHovered': string,
  +'borderRadiusMd': string,
  +'borderRadiusSm': string,
  +'borderSelected': string,
  +'check': string,
  +'checked': string,
  +'disabled': string,
  +'enabled': string,
  +'input': string,
  +'inputEnabled': string,
  +'readOnly': string,
  +'sizeMd': string,
  +'sizeSm': string,
|};
