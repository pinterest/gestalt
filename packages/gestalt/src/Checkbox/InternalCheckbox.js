// @flow strict
import {
  type AbstractComponent,
  forwardRef,
  type Node as ReactNode,
  useEffect,
  useImperativeHandle,
  useRef,
  useState,
} from 'react';
import classnames from 'classnames';
import styles from './InternalCheckbox.css';
import Box from '../Box';
import focusStyles from '../Focus.css';
import Icon from '../Icon';
import Label from '../Label';
import FormErrorMessage from '../shared/FormErrorMessage';
import FormHelperText from '../shared/FormHelperText';
import Text from '../Text';
import useFocusVisible from '../useFocusVisible';

type Props = {
  checked?: boolean,
  disabled?: boolean,
  errorMessage?: string,
  helperText?: string,
  id: string,
  image?: ReactNode,
  indeterminate?: boolean,
  label?: string,
  labelDisplay?: 'visible' | 'hidden',
  name?: string,
  onChange?: ({
    event: SyntheticInputEvent<HTMLInputElement>,
    checked: boolean,
  }) => void,
  onClick?: ({
    event: SyntheticInputEvent<HTMLInputElement>,
    checked: boolean,
  }) => void,
  /**
   * Make the checkbox readonly. Interactivity is disabled, but it can be used as a visual indicator. Click handler events are also disabled
   */
  readOnly?: boolean,
  ref?: Element<'input'>, // eslint-disable-line react/no-unused-prop-types
  size?: 'sm' | 'md',
  style?: { borderColor?: string, backgroundColor?: string },
};

const InternalCheckboxWithForwardRef: AbstractComponent<Props, HTMLInputElement> = forwardRef<
  Props,
  HTMLInputElement,
>(function Checkbox(
  {
    checked = false,
    disabled = false,
    errorMessage,
    helperText,
    id,
    image,
    indeterminate = false,
    label,
    labelDisplay = 'visible',
    name,
    onChange,
    onClick,
    readOnly = false,
    size = 'md',
    style,
  }: Props,
  ref,
): ReactNode {
  const innerRef = useRef<null | HTMLInputElement>(null);
  // When using both forwardRef and innerRef, React.useimperativehandle() allows a parent component
  // that renders <Checkbox ref={inputRef} /> to call inputRef.current.focus()
  useImperativeHandle(ref, () => innerRef.current);

  const [focused, setFocused] = useState(false);
  const [hovered, setHover] = useState(false);

  useEffect(() => {
    if (innerRef && innerRef.current) {
      innerRef.current.indeterminate = indeterminate;
    }
  }, [indeterminate]);

  const handleChange = (event: SyntheticInputEvent<HTMLInputElement>) => {
    if (onChange) {
      onChange({ event, checked: event.target.checked });
    }
  };

  const handleClick = (event: SyntheticInputEvent<HTMLInputElement>) => {
    if (onClick) {
      onClick({ event, checked: event.currentTarget.checked });
    }
  };

  const handleHover = () => {
    setHover(!hovered);
  };

  let bgStyle = styles.enabled;
  if (disabled) {
    bgStyle = styles.disabled;
  } else if (checked || indeterminate) {
    bgStyle = styles.checked;
  }

  let borderStyle = styles.border;
  if (disabled) {
    borderStyle = styles.borderDisabled;
  } else if (!disabled && (checked || indeterminate)) {
    borderStyle = styles.borderSelected;
  } else if (errorMessage) {
    borderStyle = styles.borderError;
  } else if (!disabled && hovered) {
    borderStyle = styles.borderHovered;
  }

  const borderRadiusStyle = size === 'sm' ? styles.borderRadiusSm : styles.borderRadiusMd;

  const styleSize = size === 'sm' ? styles.sizeSm : styles.sizeMd;

  const { isFocusVisible } = useFocusVisible();

  let ariaDescribedby;

  if (errorMessage) {
    ariaDescribedby = `${id}-error`;
  }

  if (label && helperText) {
    ariaDescribedby = `${id}-helperText`;
  }

  return (
    <Box>
      <Box alignItems="start" display="flex" justifyContent="start" marginEnd={-1} marginStart={-1}>
        <Box paddingX={1} position="relative">
          <input
            // checking for "focused" is not required by screenreaders but it prevents a11y integration tests to complain about missing label, as aria-describedby seems to shadow label in tests though it's a W3 accepeted pattern https://www.w3.org/TR/WCAG20-TECHS/ARIA1.html
            ref={innerRef}
            aria-describedby={focused ? ariaDescribedby : undefined}
            aria-hidden={readOnly ? true : undefined}
            aria-invalid={errorMessage ? 'true' : 'false'}
            checked={checked}
            className={classnames(styles.input, styleSize, {
              [styles.inputEnabled]: !disabled,
              [styles.readOnly]: readOnly,
            })}
            disabled={readOnly || disabled}
            id={id}
            name={name}
            onBlur={() => setFocused(false)}
            onChange={handleChange}
            onClick={handleClick}
            onFocus={() => setFocused(true)}
            onMouseEnter={handleHover}
            onMouseLeave={handleHover}
            type="checkbox"
          />
          <div
            className={classnames(
              bgStyle,
              borderStyle,
              borderRadiusStyle,
              styleSize,
              styles.check,
              {
                [focusStyles.accessibilityOutlineFocus]: focused && isFocusVisible,
              },
            )}
            style={style}
          >
            {(checked || indeterminate) && (
              <Icon
                accessibilityLabel=""
                color="inverse"
                icon={indeterminate ? 'dash' : 'check'}
                size={size === 'sm' ? 8 : 12}
              />
            )}
          </div>
        </Box>
        {Boolean(image) && <Box paddingX={1}>{image}</Box>}
        {label && (
          <Box
            dangerouslySetInlineStyle={{
              __style: { marginTop: size === 'md' ? '2px' : '-1px' },
            }}
            //  marginTop: '-1px'/'2px' is needed to  visually align the label text & radiobutton input
            display={labelDisplay === 'hidden' ? 'visuallyHidden' : 'block'}
          >
            <Label htmlFor={id}>
              <Box paddingX={1}>
                <Text color={disabled ? 'subtle' : undefined} size={size === 'sm' ? '200' : '300'}>
                  {label}
                </Text>
              </Box>
            </Label>
            <Box paddingX={1}>
              {helperText && !errorMessage ? (
                <FormHelperText id={`${id}-helperText`} size={size} text={helperText} />
              ) : null}
              {errorMessage ? (
                <FormErrorMessage id={`${id}-error`} size={size} text={errorMessage} />
              ) : null}
            </Box>
          </Box>
        )}
      </Box>
    </Box>
  );
});

InternalCheckboxWithForwardRef.displayName = 'InternalCheckbox';

export default InternalCheckboxWithForwardRef;
