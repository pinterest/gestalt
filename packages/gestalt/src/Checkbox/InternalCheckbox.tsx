import { forwardRef, ReactElement, ReactNode, useEffect, useImperativeHandle, useRef } from 'react';
import classnames from 'classnames';
import styles from './InternalCheckbox.css';
import Box from '../Box';
import focusStyles from '../Focus.css';
import Icon from '../Icon';
import Label from '../Label';
import FormErrorMessage from '../sharedSubcomponents/FormErrorMessage';
import FormHelperText from '../sharedSubcomponents/FormHelperText';
import Text from '../Text';
import useFocusVisible from '../useFocusVisible';
import useInteractiveStates from '../utils/useInteractiveStates';

type Props = {
  checked?: boolean;
  disabled?: boolean;
  errorMessage?: string;
  helperText?: string;
  id: string;
  image?: ReactNode;
  indeterminate?: boolean;
  label?: string;
  labelDisplay?: 'visible' | 'hidden';
  name?: string;
  onChange?: (arg1: { event: React.ChangeEvent<HTMLInputElement>; checked: boolean }) => void;
  onClick?: (arg1: { event: React.ChangeEvent<HTMLInputElement>; checked: boolean }) => void;
  /**
   * Make the checkbox readonly. Interactivity is disabled, but it can be used as a visual indicator. Click handler events are also disabled
   */
  readOnly?: boolean;
  ref?: ReactElement; // eslint-disable-line react/no-unused-prop-types,
  size?: 'sm' | 'md';
  style?: {
    borderColor?: string;
    backgroundColor?: string;
  };
};

const InternalCheckboxWithForwardRef = forwardRef<HTMLInputElement, Props>(function Checkbox(
  {
    checked = false,
    disabled = false,
    errorMessage,
    helperText,
    id,
    image,
    indeterminate = false,
    label,
    labelDisplay = 'visible',
    name,
    onChange,
    onClick,
    readOnly = false,
    size = 'md',
    style,
  }: Props,
  ref,
) {
  const innerRef = useRef<null | HTMLInputElement>(null);
  // When using both forwardRef and innerRef, React.useimperativehandle() allows a parent component
  // that renders <Checkbox ref={inputRef} /> to call inputRef.current.focus()
  // @ts-expect-error - TS2322 - Type 'HTMLInputElement | null' is not assignable to type 'HTMLInputElement'.
  useImperativeHandle(ref, () => innerRef.current);

  useEffect(() => {
    if (innerRef && innerRef.current) {
      innerRef.current.indeterminate = indeterminate;
    }
  }, [indeterminate]);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (onChange) {
      onChange({ event, checked: event.target.checked });
    }
  };

  const handleClick = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (onClick) {
      onClick({ event, checked: event.currentTarget.checked });
    }
  };

  const {
    handleOnMouseEnter,
    handleOnMouseLeave,
    handleOnBlur,
    handleOnFocus,
    isFocused,
    isHovered,
  } = useInteractiveStates();
  const { isFocusVisible } = useFocusVisible();

  let ariaDescribedby;

  if (errorMessage) {
    ariaDescribedby = `${id}-error`;
  }

  if (label && helperText) {
    ariaDescribedby = `${id}-helperText`;
  }

  let bgStyle = styles.enabled;
  if (disabled) {
    bgStyle = styles.disabled;
  } else if (checked || indeterminate) {
    bgStyle = styles.checked;
  }

  let borderStyle = styles.border;
  if (disabled) {
    borderStyle = styles.borderDisabled;
  } else if (!disabled && (checked || indeterminate)) {
    borderStyle = styles.borderSelected;
  } else if (errorMessage) {
    borderStyle = styles.borderError;
  } else if (!disabled && isHovered) {
    borderStyle = styles.borderHovered;
  }

  const borderRadiusStyle = size === 'sm' ? styles.borderRadiusSm : styles.borderRadiusMd;

  const styleSize = size === 'sm' ? styles.sizeSm : styles.sizeMd;

  const divStyles = classnames(bgStyle, borderStyle, borderRadiusStyle, styleSize, styles.check, {
    [focusStyles.accessibilityOutlineFocus]: isFocused && isFocusVisible,
  });

  const inputStyles = classnames(styles.input, styleSize, {
    [styles.inputEnabled]: !disabled,
    [styles.readOnly]: readOnly,
  });

  return (
    <Box>
      <Box alignItems="start" display="flex" justifyContent="start" marginEnd={-1} marginStart={-1}>
        <Box paddingX={1} position="relative">
          <input
            // checking for "focused" is not required by screenreaders but it prevents a11y integration tests to complain about missing label, as aria-describedby seems to shadow label in tests though it's a W3 accepeted pattern https://www.w3.org/TR/WCAG20-TECHS/ARIA1.html
            ref={innerRef}
            aria-describedby={isFocused ? ariaDescribedby : undefined}
            aria-hidden={readOnly ? true : undefined}
            aria-invalid={errorMessage ? 'true' : 'false'}
            checked={checked}
            className={inputStyles}
            disabled={readOnly || disabled}
            id={id}
            name={name}
            onBlur={handleOnBlur}
            onChange={handleChange}
            // @ts-expect-error - TS2322 - Type '(event: React.ChangeEvent<HTMLInputElement>) => void' is not assignable to type 'MouseEventHandler<HTMLInputElement>'.
            onClick={handleClick}
            onFocus={handleOnFocus}
            onMouseEnter={handleOnMouseEnter}
            onMouseLeave={handleOnMouseLeave}
            type="checkbox"
          />
          <div className={divStyles} style={style}>
            {(checked || indeterminate) && (
              <Icon
                accessibilityLabel=""
                color="inverse"
                icon={indeterminate ? 'dash' : 'check'}
                size={size === 'sm' ? 8 : 12}
              />
            )}
          </div>
        </Box>
        {Boolean(image) && <Box paddingX={1}>{image}</Box>}
        {label && (
          <Box
            dangerouslySetInlineStyle={{
              __style: { marginTop: size === 'md' ? '2px' : '-1px' },
            }}
            //  marginTop: '-1px'/'2px' is needed to  visually align the label text & radiobutton input
            display={labelDisplay === 'hidden' ? 'visuallyHidden' : 'block'}
          >
            <Label htmlFor={id}>
              <Box paddingX={1}>
                <Text color={disabled ? 'subtle' : undefined} size={size === 'sm' ? '200' : '300'}>
                  {label}
                </Text>
              </Box>
            </Label>
            <Box paddingX={1}>
              {helperText && !errorMessage ? (
                <FormHelperText id={`${id}-helperText`} size={size} text={helperText} />
              ) : null}
              {errorMessage ? (
                <FormErrorMessage id={`${id}-error`} size={size} text={errorMessage} />
              ) : null}
            </Box>
          </Box>
        )}
      </Box>
    </Box>
  );
});

InternalCheckboxWithForwardRef.displayName = 'InternalCheckbox';

export default InternalCheckboxWithForwardRef;
