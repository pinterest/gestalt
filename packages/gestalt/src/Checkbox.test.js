// @flow
import React from 'react';
import { create } from 'react-test-renderer';
import Checkbox from './Checkbox.js';

test('Checkbox', () => {
  const tree = create(<Checkbox id="id" onChange={() => {}} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Checkbox checked', () => {
  const tree = create(
    <Checkbox id="id" onChange={() => {}} checked />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Checkbox indeterminate', () => {
  const tree = create(
    <Checkbox id="id" onChange={() => {}} indeterminate />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Checkbox small', () => {
  const tree = create(
    <Checkbox size="sm" id="id" onChange={() => {}} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Checkbox disabled', () => {
  const tree = create(
    <Checkbox disabled size="sm" id="id" onChange={() => {}} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Checkbox disabled & checked', () => {
  const tree = create(
    <Checkbox disabled checked size="sm" id="id" onChange={() => {}} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Checkbox with error', () => {
  const tree = create(
    <Checkbox hasError size="sm" id="id" onChange={() => {}} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
