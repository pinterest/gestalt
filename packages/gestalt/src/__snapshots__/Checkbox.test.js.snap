// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox 1`] = `
<div
  className="box relative"
>
  <input
    checked={false}
    className="input inputEnabled inputMd"
    disabled={false}
    id="id"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    type="checkbox"
  />
  <div
    className="border check whiteBg checkEnabled checkMd"
  />
</div>
`;

exports[`Checkbox checked 1`] = `
<div
  className="box relative"
>
  <input
    checked={true}
    className="input inputEnabled inputMd"
    disabled={false}
    id="id"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    type="checkbox"
  />
  <div
    className="borderDark check darkGrayBg checkEnabled checkMd"
  >
    <svg
      aria-hidden={true}
      aria-label=""
      className="icon white iconBlock"
      height={12}
      role="img"
      viewBox="0 0 24 24"
      width={12}
    >
      <path
        d="test-file-stub"
      />
    </svg>
  </div>
</div>
`;

exports[`Checkbox disabled & checked 1`] = `
<div
  className="box relative"
>
  <input
    checked={true}
    className="input inputSm"
    disabled={true}
    id="id"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    type="checkbox"
  />
  <div
    className="border check grayBg checkSm"
  >
    <svg
      aria-hidden={true}
      aria-label=""
      className="icon white iconBlock"
      height={8}
      role="img"
      viewBox="0 0 24 24"
      width={8}
    >
      <path
        d="test-file-stub"
      />
    </svg>
  </div>
</div>
`;

exports[`Checkbox disabled 1`] = `
<div
  className="box relative"
>
  <input
    checked={false}
    className="input inputSm"
    disabled={true}
    id="id"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    type="checkbox"
  />
  <div
    className="border check lightGrayBg checkSm"
  />
</div>
`;

exports[`Checkbox indeterminate 1`] = `
<div
  className="box relative"
>
  <input
    checked={false}
    className="input inputEnabled indeterminate inputMd"
    disabled={false}
    id="id"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    type="checkbox"
  />
  <div
    className="borderDark check darkGrayBg checkEnabled checkMd"
  >
    <svg
      aria-hidden={true}
      aria-label=""
      className="icon white iconBlock"
      height={12}
      role="img"
      viewBox="0 0 24 24"
      width={12}
    >
      <path
        d="test-file-stub"
      />
    </svg>
  </div>
</div>
`;

exports[`Checkbox small 1`] = `
<div
  className="box relative"
>
  <input
    checked={false}
    className="input inputEnabled inputSm"
    disabled={false}
    id="id"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    type="checkbox"
  />
  <div
    className="border check whiteBg checkEnabled checkSm"
  />
</div>
`;

exports[`Checkbox with error 1`] = `
<div
  className="box relative"
>
  <input
    checked={false}
    className="input inputEnabled inputSm"
    disabled={false}
    id="id"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    type="checkbox"
  />
  <div
    className="borderError check whiteBg checkEnabled checkSm"
  />
</div>
`;
